network/src/protocols/rpc/mod.rs

```diff
@@ -169,85 +182,197 @@ pub struct Rpc<TSubstream> {
+                        if inbound_rpc_tasks.len() as u32 == self.max_concurrent_inbound_rpcs {
+                            // Log any errors.
+                            counters::LIBRA_NETWORK_RPC_MESSAGES
+                                .with_label_values(&["response", "declined"])
+                                .inc();
+                            warn!(
+                                "Pending inbound RPCs are at limit ({}). Not processing new inbound rpc requests",
+                                self.max_concurrent_inbound_rpcs
+                            );
+                            return;
+                        }
+                        let timeout = self.inbound_rpc_timeout;
+                        // Handle request with timeout.
+                        let f = async move {
+                            if let Err(err) = tokio::time::timeout(
+                                timeout,
+                                handle_inbound_request(notification_tx, request, peer_handle),
+                            )
+                            .map_err(Into::<RpcError>::into)
+                            .map(|r| r.and_then(|x| x))
+                            .await
+                            {
+                                // Log any errors.
+                                counters::LIBRA_NETWORK_RPC_MESSAGES
+                                    .with_label_values(&["response", "failed"])
+                                    .inc();
+                                warn!(
+                                    "Error handling inbound rpc request from {}: {:?}",
+                                    peer_id_str, err
+                                );
+                            }
+                        };
+                        inbound_rpc_tasks.push(f.boxed());
+                    }
+                    _ => {
+                        error!("Received non-RPC message from Peer actor: {:?}", message);
+                    }
+                }
+            }
+            notif => debug_assert!(
+                false,
+                "Received unexpected event from Peer: {:?}, expected NewMessage",
+                notif
+            ),
+        }
+    }
+
+    /// Handle an outbound rpc request.
+    ///
+    /// Cancellation is done by the client dropping the receiver side of the [`req.res_tx`]
+    /// oneshot channel. If the request is canceled, the rpc future is dropped and the request is
+    /// cancelled. Currently, we don't send a cancellation message to the remote peer.
+    ///
+    /// [`req.res_tx`]: OutboundRpcRequest::res_tx
+    async fn handle_outbound_rpc(
+        &mut self,
+        req: OutboundRpcRequest,
+        outbound_rpc_tasks: &mut FuturesUnordered<BoxFuture<'static, RequestId>>,
+    ) {
+        // If we already have too many pending RPCs, return error immediately.
+        if outbound_rpc_tasks.len() as u32 == self.max_concurrent_outbound_rpcs {
```

Comment:
Is there a reason to have a limit on the number of concurrent rpcs? I would imagine the maximum would be such that we haven't exhausted the id space?

Reply:
I personally debated keeping this as well - overall, I think it's good to have reasonable limits, since each RPC does spawn off a new task which would cause contention on the underlying connection. This is especially important for inbound RPCs, to get defense in depth. Another defense I wanted to add was an overall rate-limit on the number of messages Peer actor accepts. This will be done in the next PR as part of this migration.

Reply:
Given that this seems to be somewhat of a safe guard and not a thoroughly evaluated value, perhaps we should leave a comment there, so if folks do hit this limit in practice, they don't feel too uncomfortable bumping it up appropriately. Sometimes when we see constants in code written by other people, we assume that they were determined with great empirical evidence, we now have in our power to avoid having future folks fall into a simple trap :P.

Reply:
yes, that's definitely a trap. Will take care of letting dear reader know about the (lack of ) empirical evidence here :D